File: MiniDOMBase/DMD/Collections/CCollection.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Anagrafica
Imports System.Net
Imports minidom.Sistema

<Serializable>
Public Class CCollection
    Inherits minidom.CReadOnlyCollection

    Public Event CollectionChanged(ByVal sender As Object, ByVal e As System.EventArgs)

    Public Sub New()
    End Sub

    Public Sub New(ByVal items As IEnumerable)
        Me.New()
        Me.AddRange(items)
    End Sub

    Public Sub Clear()
        Dim e As New System.EventArgs
        Me.OnClear()
        Me.InternalClear()
        Me.OnClearComplete()
        Me.OnCollectionChanged(e)
    End Sub

    Public Overrides ReadOnly Property IsSynchronized As Boolean
        Get
            Return False
        End Get
    End Property

    Protected Overridable Sub OnClear()

    End Sub

    Protected Overridable Sub OnClearComplete()

    End Sub

    Protected Overridable Sub OnCollectionChanged(ByVal e As System.EventArgs)
        RaiseEvent CollectionChanged(Me, e)
    End Sub

    Public Overridable Property Capacity As Integer
        Get
            Return DirectCast(Me.m_List, System.Collections.ArrayList).Capacity
        End Get
        Set(value As Integer)
            DirectCast(Me.m_List, System.Collections.ArrayList).Capacity = value
        End Set
    End Property

    Private Sub InternalMerge(ByVal col As CReadOnlyCollection)
        Dim tmpa() As Object
        Dim tmpb() As Object
        Dim tmpc() As Object

        If (col.Count > 0) Then
            tmpb = col.ToArray
            Arrays.Sort(tmpb, 0, col.Count, Me.Comparer)
            If Me.Count = 0 Then
                tmpc = tmpb
            Else
                tmpa = Me.ToArray
                Arrays.Sort(tmpa, 0, Me.Count, Me.Comparer)
                tmpc = Arrays.Merge(tmpa, 0, Me.Count, tmpb, 0, col.Count, Me.Comparer)
            End If
            Me.InternalClear()
            Me.AddRange(tmpc)
        End If
    End Sub

    Public Overridable Sub Merge(ByVal col As CReadOnlyCollection)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Me.InternalMerge(col)
            End SyncLock
        Else
            Me.InternalMerge(col)
        End If
    End Sub


    Public Sub Add(ByVal item As Object)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Dim index As Integer = Me.Count
                Me.OnInsert(index, item)
                Me.InternalAdd(item)
                If Me.Sorted Then Me.Sort()
                Me.OnInsertComplete(index, item)
            End SyncLock
        Else
            Dim index As Integer = Me.Count
            Me.OnInsert(index, item)
            Me.InternalAdd(item)
            If Me.Sorted Then Me.Sort()
            Me.OnInsertComplete(index, item)
        End If
    End Sub

    Public Sub Insert(ByVal index As Integer, ByVal item As Object)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Me.OnInsert(index, item)
                Me.InternalInsert(index, item)
                If Me.Sorted Then Me.Sort()
                Me.OnInsertComplete(index, item)
            End SyncLock
        Else
            Me.OnInsert(index, item)
            Me.InternalInsert(index, item)
            If Me.Sorted Then Me.Sort()
            Me.OnInsertComplete(index, item)
        End If
    End Sub

    Default Public Shadows Property Item(ByVal index As Integer) As Object
        Get
            If Me.IsSynchronized Then
                SyncLock Me.SyncRoot
                    If (index < 0 OrElse index >= Me.Count) Then Throw New IndexOutOfRangeException()
                    Return MyBase.InternalGet(index)
                End SyncLock
            Else
                If (index < 0 OrElse index >= Me.Count) Then Throw New IndexOutOfRangeException()
                Return MyBase.InternalGet(index)
            End If
        End Get
        Set(value As Object)
            If Me.IsSynchronized Then
                SyncLock Me.SyncRoot
                    If (index < 0 OrElse index >= Me.Count) Then Throw New IndexOutOfRangeException()
                    Dim oldValue As Object = Me.InternalGet(index)
                    Me.OnSet(index, oldValue, value)
                    Me.InternalSet(index, value)
                    Me.OnSetComplete(index, oldValue, value)
                End SyncLock
            Else
                If (index < 0 OrElse index >= Me.Count) Then Throw New IndexOutOfRangeException()
                Dim oldValue As Object = Me.InternalGet(index)
                Me.OnSet(index, oldValue, value)
                Me.InternalSet(index, value)
                Me.OnSetComplete(index, oldValue, value)
            End If
        End Set
    End Property

    Protected Overridable Sub InternalSet(ByVal index As Integer, ByVal value As Object)
        DirectCast(Me.m_List, System.Collections.ArrayList).Item(index) = value
    End Sub

    Public Overridable Sub RemoveAt(ByVal index As Integer)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Dim value As Object = Me.InternalGet(index)
                Me.OnRemove(index, value)
                Me.InternalRemoveAt(index)
                Me.OnRemoveComplete(index, value)
                Me.OnCollectionChanged(New System.EventArgs)
            End SyncLock
        Else
            Dim value As Object = Me.InternalGet(index)
            Me.OnRemove(index, value)
            Me.InternalRemoveAt(index)
            Me.OnRemoveComplete(index, value)
            Me.OnCollectionChanged(New System.EventArgs)
        End If
    End Sub

    Protected Overridable Sub OnRemove(index As Integer, value As Object)

    End Sub

    Protected Overridable Sub OnRemoveComplete(index As Integer, value As Object)

    End Sub


    Protected Overrides Sub InternalRemoveAt(ByVal index As Integer)
        DirectCast(Me.m_List, System.Collections.ArrayList).RemoveAt(index)
    End Sub

    Public Sub Remove(ByVal item As Object)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Me.RemoveAt(Me.IndexOf(item))
            End SyncLock
        Else
            Me.RemoveAt(Me.IndexOf(item))
        End If
    End Sub

    Public Sub AddRange(ByVal items As IEnumerable)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                For Each item As Object In items
                    Dim index As Integer = Me.Count
                    Me.OnInsert(index, item)
                    Me.InternalAdd(item)
                    Me.OnInsertComplete(index, item)
                Next
                If Me.Sorted Then Me.Sort()
            End SyncLock
        Else
            For Each item As Object In items
                Dim index As Integer = Me.Count
                Me.OnInsert(index, Item)
                Me.InternalAdd(Item)
                Me.OnInsertComplete(index, item)
            Next
            If Me.Sorted Then Me.Sort()
        End If
    End Sub

    Protected Overrides Sub OnInsertComplete(index As Integer, value As Object)
        MyBase.OnInsertComplete(index, value)
        Me.OnCollectionChanged(New System.EventArgs)
    End Sub

   
    Protected Overrides Sub OnSetComplete(index As Integer, oldValue As Object, newValue As Object)
        MyBase.OnSetComplete(index, oldValue, newValue)
        Me.OnCollectionChanged(New System.EventArgs)
    End Sub

    Public Overridable Function IsChanged() As Boolean
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                For Each o As Object In Me
                    If (DBUtils.IsChanged(o)) Then Return True
                Next
                Return False
            End SyncLock
        Else
            For Each o As Object In Me
                If (DBUtils.IsChanged(o)) Then Return True
            Next
            Return False
        End If
    End Function

    Public Overridable Sub SetChanged(ByVal value As Boolean)
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                For Each o As Object In Me
                    DBUtils.SetChanged(o, value)
                Next
            End SyncLock
        Else
            For Each o As Object In Me
                DBUtils.SetChanged(o, value)
            Next
        End If
    End Sub

    Public Overridable Function IntersectWith(ByVal col As CCollection) As CCollection
        If Me.IsSynchronized Then
            SyncLock Me.SyncRoot
                Dim ret As CCollection = Types.CreateInstance(Me.GetType)
                For Each a As Object In Me
                    For Each b As Object In col
                        If (Arrays.Compare(a, b, Me.Comparer) = 0) Then
                            ret.Add(a)
                        End If
                    Next
                Next
                Return ret
            End SyncLock
        Else
            Dim ret As CCollection = Types.CreateInstance(Me.GetType)
            For Each a As Object In Me
                For Each b As Object In col
                    If (Arrays.Compare(a, b, Me.Comparer) = 0) Then
                        ret.Add(a)
                    End If
                Next
            Next
            Return ret
        End If
    End Function

End Class

File: MiniDOMBase/DMD/Collections/CCollection.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;

namespace minidom
{

    /// <summary>
    /// Handler dell'evento Changed della Collezione
    /// </summary>
    /// <param name="sender"></param>
    /// <param name="e"></param>
    public delegate void CollectionChangedEventHandler(object sender, EventArgs e);

    /// <summary>
    /// Collezione di oggetti
    /// </summary>
    [Serializable]
    public class CCollection 
        : CReadOnlyCollection
    {

        /// <summary>
        /// Evento generato quando la collezione viene modificata
        /// </summary>
        public event CollectionChangedEventHandler CollectionChanged;

        
        /// <summary>
        /// Costruttore
        /// </summary>
        public CCollection()
        {
        }

        /// <summary>
        /// Costruttore
        /// </summary>
        /// <param name="items"></param>
        public CCollection(IEnumerable items) : this()
        {
            AddRange(items);
        }

        /// <summary>
        /// Rimuove tutti gli elementi dalla collezione
        /// </summary>
        public void Clear()
        {
            var e = new EventArgs();
            this.OnClear();
            InternalClear();
            this.OnClearComplete();
            OnCollectionChanged(e);
        }

        /// <summary>
        /// OnClear
        /// </summary>
        protected virtual void OnClear()
        {

        }

        /// <summary>
        /// OnClearComplete
        /// </summary>
        protected virtual void OnClearComplete()
        {

        }

        /// <summary>
        /// Genera l'evento CollectionChanged
        /// </summary>
        /// <param name="e"></param>
        protected virtual void OnCollectionChanged(EventArgs e)
        {
            CollectionChanged?.Invoke(this, e);
        }

        /// <summary>
        /// Restituisce o imposta la capacità della collezione
        /// </summary>
        public virtual int Capacity
        {
            get
            {
                return ((ArrayList)m_List).Capacity;
            }

            set
            {
                ((ArrayList)m_List).Capacity = value;
            }
        }

        /// <summary>
        /// Unisce la collezione
        /// </summary>
        /// <param name="col"></param>
        private void InternalMerge(CReadOnlyCollection col)
        {
            object[] tmpa;
            object[] tmpb;
            object[] tmpc;
            if (col.Count > 0)
            {
                tmpb = col.ToArray();
                DMD.Arrays.Sort(tmpb, 0, col.Count, Comparer);
                if (Count == 0)
                {
                    tmpc = tmpb;
                }
                else
                {
                    tmpa = ToArray();
                    DMD.Arrays.Sort(tmpa, 0, Count, Comparer);
                    tmpc = DMD.Arrays.Merge(tmpa, 0, Count, tmpb, 0, col.Count, Comparer);
                }

                InternalClear();
                AddRange(tmpc);
            }
        }

        /// <summary>
        /// Effettua la UNION delle due collezioni
        /// </summary>
        /// <param name="col"></param>
        public virtual void Merge(CReadOnlyCollection col)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                    InternalMerge(col);
            }
            else
            {
                InternalMerge(col);
            }
        }

        /// <summary>
        /// Aggiunge l'elemto alla collezione
        /// </summary>
        /// <param name="item"></param>
        public void Add(object item)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    int index = Count;
                    OnInsert(index, item);
                    InternalAdd(item);
                    if (Sorted)
                        Sort();
                    OnInsertComplete(index, item);
                }
            }
            else
            {
                int index = Count;
                OnInsert(index, item);
                InternalAdd(item);
                if (Sorted)
                    Sort();
                OnInsertComplete(index, item);
            }
        }

        /// <summary>
        /// Inserisce l'elemento nella collezione
        /// </summary>
        /// <param name="index"></param>
        /// <param name="item"></param>
        public void Insert(int index, object item)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    OnInsert(index, item);
                    InternalInsert(index, item);
                    if (Sorted)
                        Sort();
                    OnInsertComplete(index, item);
                }
            }
            else
            {
                OnInsert(index, item);
                InternalInsert(index, item);
                if (Sorted)
                    Sort();
                OnInsertComplete(index, item);
            }
        }

        /// <summary>
        /// Restituisce o imposta l'elemento nella posizione specifica
        /// </summary>
        /// <param name="index"></param>
        /// <returns></returns>
        public new object this[int index]
        {
            get
            {
                if (IsSynchronized)
                {
                    lock (SyncRoot)
                    {
                        if (index < 0 || index >= Count)
                            throw new IndexOutOfRangeException();
                        return base.InternalGet(index);
                    }
                }
                else
                {
                    if (index < 0 || index >= Count)
                        throw new IndexOutOfRangeException();
                    return base.InternalGet(index);
                }
            }

            set
            {
                if (IsSynchronized)
                {
                    lock (SyncRoot)
                    {
                        if (index < 0 || index >= Count)
                            throw new IndexOutOfRangeException();
                        var oldValue = InternalGet(index);
                        OnSet(index, oldValue, value);
                        InternalSet(index, value);
                        OnSetComplete(index, oldValue, value);
                    }
                }
                else
                {
                    if (index < 0 || index >= Count)
                        throw new IndexOutOfRangeException();
                    var oldValue = InternalGet(index);
                    OnSet(index, oldValue, value);
                    InternalSet(index, value);
                    OnSetComplete(index, oldValue, value);
                }
            }
        }

        

        public virtual void RemoveAt(int index)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    var value = InternalGet(index);
                    OnRemove(index, value);
                    InternalRemoveAt(index);
                    OnRemoveComplete(index, value);
                    OnCollectionChanged(new EventArgs());
                }
            }
            else
            {
                var value = InternalGet(index);
                OnRemove(index, value);
                InternalRemoveAt(index);
                OnRemoveComplete(index, value);
                OnCollectionChanged(new EventArgs());
            }
        }

        protected virtual void OnRemove(int index, object value)
        {
        }

        protected virtual void OnRemoveComplete(int index, object value)
        {
        }

        protected override void InternalRemoveAt(int index)
        {
            ((ArrayList)m_List).RemoveAt(index);
        }

        public void Remove(object item)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                    RemoveAt(IndexOf(item));
            }
            else
            {
                RemoveAt(IndexOf(item));
            }
        }

        public void AddRange(IEnumerable items)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    foreach (object item in items)
                    {
                        int index = Count;
                        OnInsert(index, item);
                        InternalAdd(item);
                        OnInsertComplete(index, item);
                    }

                    if (Sorted)
                        Sort();
                }
            }
            else
            {
                foreach (object item in items)
                {
                    int index = Count;
                    OnInsert(index, item);
                    InternalAdd(item);
                    OnInsertComplete(index, item);
                }

                if (Sorted)
                    Sort();
            }
        }

        protected override void OnInsertComplete(int index, object value)
        {
            base.OnInsertComplete(index, value);
            OnCollectionChanged(new EventArgs());
        }

        protected override void OnSetComplete(int index, object oldValue, object newValue)
        {
            base.OnSetComplete(index, oldValue, newValue);
            OnCollectionChanged(new EventArgs());
        }

        public virtual bool IsChanged()
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    foreach (object o in this)
                    {
                        if (Databases.DBUtils.IsChanged(o))
                            return true;
                    }

                    return false;
                }
            }
            else
            {
                foreach (object o in this)
                {
                    if (Databases.DBUtils.IsChanged(o))
                        return true;
                }

                return false;
            }
        }

        public virtual void SetChanged(bool value)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    foreach (object o in this)
                        Databases.DBUtils.SetChanged(o, value);
                }
            }
            else
            {
                foreach (object o in this)
                    Databases.DBUtils.SetChanged(o, value);
            }
        }

        public virtual CCollection IntersectWith(CCollection col)
        {
            if (IsSynchronized)
            {
                lock (SyncRoot)
                {
                    CCollection ret = (CCollection)Sistema.Types.CreateInstance(GetType());
                    foreach (object a in this)
                    {
                        foreach (object b in col)
                        {
                            if (DMD.Arrays.Compare(a, b, Comparer) == 0)
                            {
                                ret.Add(a);
                            }
                        }
                    }

                    return ret;
                }
            }
            else
            {
                CCollection ret = (CCollection)Sistema.Types.CreateInstance(GetType());
                foreach (object a in this)
                {
                    foreach (object b in col)
                    {
                        if (DMD.Arrays.Compare(a, b, Comparer) == 0)
                        {
                            ret.Add(a);
                        }
                    }
                }

                return ret;
            }
        }
    }
}